### PR TITLE
Feature sum

### PR DESCRIPTION
**Pull request by codepadma**

### CODE REVIEW

File: client/js/table-view.js [hunk 1]
```diff
@@ -18,11 +18,13 @@ class TableView {
   	this.headerRowEle = document.querySelector('THEAD TR');
```
Comment (speschellk) [inline client/js/table-view.js]: Use consistent indentation to make your code more easily readable.

File: client/js/table-view.js [hunk 2]
```diff
@@ -18,11 +18,13 @@ class TableView {
...
     this.bodyEle = document.querySelector('TBODY');
     this.formulaBarEle = document.querySelector('#formula-bar');
+    this.footEle = document.querySelector('TFOOT');
   }
 
   renderTable() {
   	this.renderTableHeader();
```
Comment (speschellk) [inline client/js/table-view.js]: Indentation issue

File: client/js/table-view.js [hunk 3]
```diff
@@ -18,11 +18,13 @@ class TableView {
   	this.headerRowEle = document.querySelector('THEAD TR');
     this.bodyEle = document.querySelector('TBODY');
```
Comment (speschellk) [inline client/js/table-view.js]: A more standard abbreviation for "element" is "el". Please switch.

File: client/js/test/table-view.test.js [hunk 4]
```diff
@@ -28,6 +28,24 @@ describe('table-view', () => {
   });
   });	
 
+  describe('table-foot', () => {
+    it('table foot has the sum of col values', () => {
```
Comment (speschellk) [inline client/js/test/table-view.test.js]: Test assertions read like a sentence, so they should begin with a verb. This one would say, "It table foot has the sum of col values". A better test assertion would say "it DISPLAYS the sum of column values in the table footer".

File: client/js/test/table-view.test.js [hunk 5]
```diff
@@ -28,6 +28,24 @@ describe('table-view', () => {
   });
   });
```
Comment (speschellk) [inline client/js/test/table-view.test.js]: Add a few more tests that test the functionality of your new feature. Especially test against non-happy paths and edge cases. For instance, “it correctly generates sum when the column includes non-number inputs”.

**Review by speschellk — CHANGES_REQUESTED** [review]

File: client/js/table-view.js (hunk 2 above)
Comment (codepadma) [inline client/js/table-view.js]:
```initDomReferences() {
    this.headerRowEl = document.querySelector('THEAD TR');
    this.bodyEl = document.querySelector('TBODY');
    this.formulaBarEl = document.querySelector('#formula-bar');
    this.footEl = document.querySelector('TFOOT');
  }```

Comment (codepadma) [inline client/js/table-view.js]: Fixed

Comment (codepadma) [inline client/js/table-view.js]: initDomReferences() {
    this.headerRowEl = document.querySelector('THEAD TR');
    this.bodyEl = document.querySelector('TBODY');
    this.formulaBarEl = document.querySelector('#formula-bar');
    this.footEl = document.querySelector('TFOOT');
  }

File: client/js/table-view.js (hunk 1 above)
Comment (codepadma) [reply client/js/table-view.js]: Fixed

File: client/js/table-view.js (hunk 3 above)
Comment (codepadma) [reply client/js/table-view.js]: Fixed.

File: client/js/test/table-view.test.js (hunk 5 above)
Comment (codepadma) [reply client/js/test/table-view.test.js]: Added edgecases and test for non-happy paths

File: client/js/test/table-view.test.js (hunk 4 above)
Comment (codepadma) [reply client/js/test/table-view.test.js]: Fixed

File: client/js/table-view.js [hunk 6]
```diff
@@ -62,13 +63,49 @@ class TableView {
...
+    }
+    docFragment.appendChild(tr);
+    this.footEl.appendChild(docFragment);
+    this.setFootColor();
+  }
+
+  calculateColSum(col) {
+    const trs = document.querySelectorAll('TBODY TR');
+    return Array.from(trs)
+      .map(function(tr) {
```
Comment (speschellk) [inline client/js/table-view.js]: This is a bulky part of the function. Could you separate it into a helper function?

File: client/js/test/fixtures/sheet-container.html [hunk 7]
```diff
@@ -9,5 +9,7 @@
     	<tr>
     	</tr>
     </tbody>
+    <tfoot>
+    </tfoot>
 	</table>
```
Comment (speschellk) [inline client/js/test/fixtures/sheet-container.html]: Careful of the indentation. All nested portions should be indented from their parents.

File: public/index.html [hunk 8]
```diff
@@ -8,14 +8,16 @@
   <div class="container">
```
Comment (speschellk) [inline public/index.html:8]: Check the indentation in this file.

**Review by speschellk — CHANGES_REQUESTED** [review]

File: client/js/table-view.js (hunk 6 above)
Comment (codepadma) [reply client/js/table-view.js]: Do you suggest moving 'calculateColSum' into one of the util files(dom-util/array-util) ?

File: client/js/test/fixtures/sheet-container.html (hunk 7 above)
Comment (codepadma) [reply client/js/test/fixtures/sheet-container.html]: Corrected indentation in all the files

File: client/js/table-view.js (hunk 6 above)
Comment (codepadma) [reply client/js/table-view.js]: Fixed! renderTableFoot decomposed and made use of helper functions

File: client/js/dom-util.js [hunk 9]
```diff
@@ -1,27 +1,31 @@
...
-  	const el = document.createElement(tagName);
+    const el = document.createElement(tagName);
     if (text) {
       el.textContent = text;
     }
     return el;
   }
 };
 
+const getTableRows = function() {
```
Comment (tim-hr) [inline client/js/dom-util.js]: Nice factoring-out of a helper function. 

A more-precise name would be `getTableBodyRows`, since there can also be rows in the header and footer, but your selector string specifies only rows under the `TBODY`.

File: client/js/table-view.js [hunk 10]
```diff
@@ -62,13 +63,52 @@ class TableView {
       }
       docFragment.appendChild(tr);
     }
-    removeChildren(this.bodyEle);
-    this.bodyEle.appendChild(docFragment);
+    removeChildren(this.bodyEl);
+    this.bodyEl.appendChild(docFragment);
+  }
+
+  renderTableFoot() {
```
Comment (tim-hr) [inline client/js/table-view.js]: even though the tag name, probably for brevity's sake, was called `TFOOT`, the name of this function, for readability's sake, should be `renderTableFooter()`.

File: client/js/table-view.js [hunk 11]
```diff
@@ -7,22 +7,23 @@ class TableView {
...
   }
 
   initDomReferences() {
-  	this.headerRowEle = document.querySelector('THEAD TR');
-    this.bodyEle = document.querySelector('TBODY');
-    this.formulaBarEle = document.querySelector('#formula-bar');
+    this.headerRowEl = document.querySelector('THEAD TR');
+    this.bodyEl = document.querySelector('TBODY');
+    this.formulaBarEl = document.querySelector('#formula-bar');
+    this.footEl = document.querySelector('TFOOT');
```
Comment (tim-hr) [inline client/js/table-view.js]: even though the tag name, probably for brevity's sake, was called `TFOOT`, the name of this property, for readability's sake, should be `footerEl`.

File: public/css/styles.css [hunk 12]
```diff
@@ -15,6 +15,14 @@ td {
   background-color: pink
 }
 
+.highlight-foot {
```
Comment (tim-hr) [inline public/css/styles.css]: `.highlight-footer`

Note that you could have inferred this from looking around you in this same code file. There's already something here called `.header`.  If you have the name "header", you should have the name "footer".

Consistency is so so so important for keeping the code comprehensible.

When multiple developers all wander around the code base doing their own thing, the result over time is a bug-prone mess.  Especially in an un-compiled language like JavaScript -- you have to be extra vigilant against typos.

File: client/js/table-view.js [hunk 13]
```diff
@@ -62,13 +63,52 @@ class TableView {
...
+
+  validateColumnVals(cellVal) {
+    return (!Number.isNaN(cellVal)) ? cellVal : '';
+  }
+
+  computeColumnSum(colVals) {
+    return colVals.reduce((sum, cellVal) => sum + cellVal);
+  }
+
+  setFootColor() {
```
Comment (tim-hr) [inline client/js/table-view.js]: `setFooterClass` -- you're over-assuming what the application of the CSS class `highlight-foot` (which should be renamed, see below) is doing.  What if it also bolds, or changes the size, or something? 

Even a small unnecessary coupling between two code locations like this should be avoided wherever possible.

Comment (tim-hr) [inline client/js/table-view.js]: Furthermore I don't think there's much benefit to factoring out this helper function. It may as well go back inline with the rest of the DOM manipulation occurring in the calling function.

Why? Because 

* it's only called in one place, 
* the body is only one line long, 
* that one line is doing a trivial amount of processing, 
* that line is completely self-describing as it is.

We don't gain any re-use or "easier to think about because the details are abstracted away" benefits from this particular helper function so I propose you just discard it.

File: client/js/table-view.js [hunk 14]
```diff
@@ -62,13 +63,52 @@ class TableView {
...
+    return getTableRows()
+      .map((tr) => {
+        const cellVal = parseInt(tr.cells[colIndex].textContent, 10); 
+        return this.validateColumnVals(cellVal);
+      })
+      .filter((cellVal) => cellVal !== '');
+  }
+
+  validateColumnVals(cellVal) {
+    return (!Number.isNaN(cellVal)) ? cellVal : '';
```
Comment (tim-hr) [inline client/js/table-view.js]: You don't need the surrounding parens, it's no big deal but less visual cruft is good.

File: client/js/test/dom-util.test.js [hunk 15]
```diff
@@ -6,16 +6,16 @@ const { removeChildren,
...
         const ele = createTH();
         expect(ele.tagName).toBe('TH');
       });
 
       it('sets the text of TH element', () => {
-      	const text = 'Table Header';
-      	const ele = createTH(text);
-      	expect(ele.textContent).toBe(text);
+        const text = 'Table Header';
+        const ele = createTH(text);
```
Comment (tim-hr) [inline client/js/test/dom-util.test.js]: Elsewhere we say `el` not `ele`. Stay consistent, please.

**Review by tim-hr — CHANGES_REQUESTED** [review]

File: client/js/dom-util.js (hunk 9 above)
Comment (codepadma) [reply client/js/dom-util.js]: Fixed the name of the function

File: client/js/table-view.js (hunk 11 above)
Comment (codepadma) [reply client/js/table-view.js]: Changed the property name to footerEl

File: client/js/table-view.js (hunk 10 above)
Comment (codepadma) [reply client/js/table-view.js]: Changed the function name to renderTableFooter

File: client/js/table-view.js (hunk 13 above)
Comment (codepadma) [reply client/js/table-view.js]: Removed the helper function and added code inline

File: public/css/styles.css (hunk 12 above)
Comment (codepadma) [reply public/css/styles.css]: Changed the class name to .highlight-footer

File: client/js/test/dom-util.test.js (hunk 15 above)
Comment (codepadma) [reply client/js/test/dom-util.test.js]: Fixed!
